### ISSUE TITLE
XBox360 not working

### ISSUE THREAD

**Issue by JarlJansen** [masmu/pulseaudio-dlna#108]
When I try to play liveaudio from linux via pulseaudio-dlna/pavucontrol I get error message from pulaseaudio-dlna like

10-16 23:15:52 pulseaudio_dlna.pulseaudio                     ERROR    The device "Xbox 360 (DLNA)" failed to play! (500)

I have seen that XBox360 is not yet tested/supported. Will the XBox360 be supported by  pulseaudio-dlna any time ?

**Comment by masmu** [comment]
There is no such thing as a list of devices I am planning to support. I try to do my best to make it work with every device I get my hands on. And that's somehow the problem. I do not own a Xbox 360 and therefore I need a lot of help from the people who is owning the device. That means typically: A process over some days to some weeks, where I:
- ask for logs
- ask for dumps
- ask you to try this
- ask you to try that

But the XBox seems to be a good device for that. It is pretty popular. Wanna go on? The first thing I need is a full _debug.log_.


```bash
pulseaudio-dlna --debug 2>&1 | tee -a debug.log
```

Could you paste it somewhere and link it here?

**Comment by JarlJansen** [comment]
Great :-)  Let's go on. Here you can find the debug.log;

http://pastebin.com/raw.php?i=dsKXDikT

**Comment by JarlJansen** [comment]
Sorry, I was wrong. XBox is really playing but it takes up to 45 seconds before playing starts. Is there any buffer that can be decreased to reduce this delay ? 
And there is another issue: When I change to another song in Spotiy the XBox stops playing.

**Comment by masmu** [comment]
Thank you for testing this yourself a bit more. 45 seconds is quite a huge delay. 
There are two explanations possible. For the first one see the topic "_There is a delay about a few seconds_" in the readme. But in that case it is more likely that the second one is correct, because no buffer should take 45 seconds to fill. How is that device connected? Weak wifi, cable?

So, I can only assume that this is because of _microsoft_'s implementation of detecting mp3 streams.
A little more explanation:
If you transcode an audio file via _lame_ there is a _mp3 header_ written to the new file *after* the transcoding finished. In that header you will find information about the used codec settings, information which must be detected by the device's decoder if those are missing. Since _pulseaudio-dlna_ streams your live audio there is no such thing as and end and the lame encoder won't send the header at all. So, if you are using _lame_ you are relying on a good and fast device decoder detection. There are some devices which can handle that really well and very fast. E.g. the _D-Link DCH-M225/E_ which is using _mplayer_ (awesome!) internally. Detection time: Nearly none.
Then I've experienced devices which do take a little longer. E.g. all the _Sonos_ ones. They need approximately 5 seconds for that. But 45 seconds is huge :wink: 

As an alternative I could play with _ffmpeg_ (can also produce _mp3_) as encoder. Perhaps it is writing the header at the beginning, never tested that.

There are other encoders which do write that header at the beginning, e.g. _sox_ (_wav_, _l16_). So, to test that, I would recommend to use those codecs and see what happens.
```
pulseaudio-dlna --encoder wav
pulseaudio-dlna --encoder l16
```

**Comment by JarlJansen** [comment]
I got an error that wav is no suitable codec for the xbox. I16 works with delay of 5 seconds and sometimes interruptions, but speed is too fast during playback. Maybe is has to do with this error:

10-21 22:28:29 pulseaudio_dlna.streamserver                   INFO     Processes of /dWRuPSJ1dWlkOjA4MTk5MDExLTM2MDUtMjAwMC0wMDAwLTdjZWQ4ZGY4OGYzNiI%3D/stream.pcm16 initialized ...
sox WARN wav: Length in output .wav header will be wrong since can't seek to fix it

**Comment by masmu** [comment]
How is your device connected to you LAN? Cable or Wifi? If it is Wifi, how is the signal strength?

Your device claims to support the following codecs:
```
http-get:*:audio/mpeg:DLNA.ORG_PN=MP3
http-get:*:audio/mpeg:DLNA.ORG_PN=MP3X
http-get:*:audio/x-ms-wma:DLNA.ORG_PN=WMABASE
http-get:*:audio/x-ms-wma:DLNA.ORG_PN=WMAFULL
http-get:*:audio/x-ms-wma:DLNA.ORG_PN=WMALSL
http-get:*:audio/x-ms-wma:DLNA.ORG_PN=WMALSL_MULT5
http-get:*:audio/x-ms-wma:DLNA.ORG_PN=WMAPRO
http-get:*:audio/L16:DLNA.ORG_PN=LPCM
http-get:*:audio/L16;rate=44100;channels=1:DLNA.ORG_PN=LPCM
http-get:*:audio/L16;rate=44100;channels=2:DLNA.ORG_PN=LPCM
http-get:*:audio/L16;rate=48000;channels=1:DLNA.ORG_PN=LPCM
http-get:*:audio/L16;rate=48000;channels=2:DLNA.ORG_PN=LPCM
http-get:*:image/jpeg:DLNA.ORG_PN=JPEG_LRG
http-get:*:image/jpeg:DLNA.ORG_PN=JPEG_MED
http-get:*:image/jpeg:DLNA.ORG_PN=JPEG_SM
http-get:*:video/mpeg:DLNA.ORG_PN=MPEG_PS_NTSC
http-get:*:video/mpeg:DLNA.ORG_PN=MPEG_TS_SD_NA_ISO
http-get:*:video/mpeg:DLNA.ORG_PN=AVC_TS_MP_SD_AAC_MULT5_ISO
http-get:*:video/mpeg:DLNA.ORG_PN=AVC_TS_MP_HD_AAC_MULT5_ISO
http-get:*:video/vnd.dlna.mpeg-tts:DLNA.ORG_PN=MPEG_TS_SD_NA
http-get:*:video/vnd.dlna.mpeg-tts:DLNA.ORG_PN=MPEG_TS_SD_NA_T
http-get:*:video/x-ms-asf:DLNA.ORG_PN=VC1_ASF_AP_L1_WMA
http-get:*:video/x-ms-asf:DLNA.ORG_PN=VC1_ASF_AP_L2_WMA
http-get:*:video/x-ms-wmv:DLNA.ORG_PN=WMVHIGH_FULL
http-get:*:video/x-ms-wmv:DLNA.ORG_PN=WMVHIGH_PRO
http-get:*:video/x-ms-wmv:DLNA.ORG_PN=WMVMED_BASE
http-get:*:video/x-ms-wmv:DLNA.ORG_PN=WMVMED_FULL
http-get:*:video/x-ms-wmv:DLNA.ORG_PN=WMVMED_PRO
http-get:*:video/x-ms-wmv:DLNA.ORG_PN=WMVSPLL_BASE
http-get:*:video/x-ms-wmv:DLNA.ORG_PN=WMVSPML_BASE
http-get:*:audio/mp4:*
http-get:*:audio/mpeg:*
http-get:*:audio/x-ms-wma:*
http-get:*:audio/L16:*
http-get:*:image/jpeg:*
http-get:*:video/avi:*
http-get:*:video/mpeg:*
http-get:*:video/vnd.dlna.mpeg-tts:*
http-get:*:video/mp4:*
http-get:*:video/quicktime:*
http-get:*:video/x-ms-asf:*
http-get:*:video/x-ms-wmv:*
microsoft.com:*:application/vnd.ms-playtoapp;target="ms-playtoapp-xboxvideo":*
microsoft.com:*:application/vnd.ms-playtoapp;target="ms-playtoapp-xboxmusic":*
```
So, that _wav_ does not work is because _microsoft_ does not support it. But that is not that much a problem. _l16_ and _wav_ are pretty similar.

Could you paste me another _debug.log_ with the _l16_ encoder? And I need the output of
```bash
pactl list sink-inputs
```
while this is running.

**Comment by JarlJansen** [comment]
I have a Netbook conncted to my Wifi-Router via Wifi 150MBit/s, my Xbox360 is also connected via Wifi 72MBit/s . That should be far enough for streaming.

Here is tdebug.log for I16:

http://pastebin.com/raw.php?i=PJzqGvwx

Here is pactl output:

http://pastebin.com/raw.php?i=B4yN7n2D

**Comment by masmu** [comment]
Thanks, the output of _pactl_ is fine, but could you do a full _debug.log_ like the last time?

```bash
pulseaudio-dlna --debug 2>&1 | tee -a debug.log
```

**Comment by masmu** [comment]
And I finally got an idea what is happening. Could you run `pulseaudio-dlna --create-device-config`one time? It will create a _devices.json_. Please paste it here, I change it here to a hopefully working example.

**Comment by JarlJansen** [comment]
Here you are:

debug.log 

http://pastebin.com/raw.php?i=CdQviM5s

**Comment by masmu** [comment]
Thanks. I think _pulseaudio-dlna_ selects the wrong sample rate or it is converted the wrong way. Could you provide the mentioned _devices.json_? It think we can make it work in 2 minutes ...

**Comment by JarlJansen** [comment]
Here is device config:

{
    "uuid:08199011-3605-2000-0000-7ced8df88f36": {
        "rules": [], 
        "flavour": "DLNA", 
        "name": "Xbox 360", 
        "codecs": [
            {
                "rules": [], 
                "bit_rate": null, 
                "identifier": "mp3", 
                "mime_type": "audio/mpeg"
            }, 
            {
                "channels": 2, 
                "rules": [], 
                "identifier": "l16", 
                "sample_rate": 48000, 
                "mime_type": "audio/L16;rate=48000;channels=2"
            }, 
            {
                "channels": 2, 
                "rules": [], 
                "identifier": "l16", 
                "sample_rate": 44100, 
                "mime_type": "audio/L16;rate=44100;channels=2"
            }, 
            {
                "channels": 1, 
                "rules": [], 
                "identifier": "l16", 
                "sample_rate": 48000, 
                "mime_type": "audio/L16;rate=48000;channels=1"
            }, 
            {
                "channels": 1, 
                "rules": [], 
                "identifier": "l16", 
                "sample_rate": 44100, 
                "mime_type": "audio/L16;rate=44100;channels=1"
            }, 
            {
                "channels": null, 
                "rules": [], 
                "identifier": "l16", 
                "sample_rate": null, 
                "mime_type": "audio/L16"
            }
        ]
    }
}

**Comment by masmu** [comment]
Ok, please replace your _devices.json_ with the following:
```json
{
    "uuid:08199011-3605-2000-0000-7ced8df88f36": {
        "rules": [], 
        "flavour": "DLNA", 
        "name": "Xbox 360", 
        "codecs": [
            {
                "channels": 2, 
                "rules": [], 
                "identifier": "l16", 
                "sample_rate": 44100, 
                "mime_type": "audio/L16;rate=44100;channels=2"
            }
        ]
    }
}```

**Comment by masmu** [comment]
Then just start _pulseaudio-dlna_ without specifying any flags. No `--encoder` and no `--bit-rate` flags at least ...

**Comment by JarlJansen** [comment]
Works fine :-) It's playing with proper speed without any interruptions. 

Thx a lot for your support.
Now pulseaudio-dlna is also XBox360 certified ;-)

**Comment by masmu** [comment]
We now know why the problem occurred. But other _Xbox 360_ users will still suffer from this. Can you help me to fix the problem for all? It will require you to install a local version of _pulseaudio-dlna_ ...

**Comment by JarlJansen** [comment]
Sure. I can do this for you.

**Comment by masmu** [comment]
Ok, are you on Ubuntu and are you familiar with _git_?

**Comment by JarlJansen** [comment]
Not yet.

**Comment by masmu** [comment]
No problem. Glad that you help me to fix this for all :wink: 

Please remove the created _devices.json_, so that the problem should reoccur.

```bash
# remove the ppa based pulseaudio-dlna version
sudo apt-get remove pulseaudio-dlna

# install all dependencies (if you installed the ppa version before this should not install anything new)
sudo apt-get install python2.7 python-pip python-setuptools python-dbus python-beautifulsoup python-docopt python-requests python-setproctitle python-gobject python-protobuf python-notify2 python-psutil python-concurrent.futures python-chardet vorbis-tools sox lame flac faac opus-tools

# install dependencies necessary for python virtual env
sudo apt-get install virtualenv python-dev git

# goto some folder, e.g. your home folder
cd ~

# clone the actual code and change to that directory
git clone https://github.com/masmu/pulseaudio-dlna.git
cd pulseaudio-dlna/

# checkout the branch
git checkout bug/xbox-360

# make the application runnable
make

# run it
bin/pulseaudio-dlna
```

I am currently working on the fix...

**Comment by JarlJansen** [comment]
done

**Comment by JarlJansen** [comment]
Is this version already including the bug-fix ?

**Comment by masmu** [comment]
Thank you! No, I am still working on it... I am not quite sure if there is a bug in microsofts _l16_ implementation or I am doing this wrong. I guess i am doing it wrong :smile:

**Comment by JarlJansen** [comment]
I removed my encoders together with pulseaudio-dlna. How to reinstall I16 on Ubuntu ?

**Comment by masmu** [comment]
```bash
sudo apt-get install python2.7 python-pip python-setuptools python-dbus python-beautifulsoup python-docopt python-requests python-setproctitle python-gobject python-protobuf python-notify2 python-psutil python-concurrent.futures python-chardet vorbis-tools sox lame flac faac opus-tools
```

_sox_ is the _wav_ and _l16_ encoder. So it should be enought 
```bash
sudo apt-get install sox
```

**Comment by JarlJansen** [comment]
Not working:

sudo apt-get -u install sox
Reading package lists... Done
Building dependency tree       
Reading state information... Done
sox is already the newest version.
The following packages were automatically installed and are no longer required:
  account-plugin-windows-live faac flac lame libao-common libao4 libavutil52
  libfaac0 libjs-jquery libjs-sphinxdoc libjs-underscore libopus0
  libpostproc52 libschroedinger-1.0-0 libswscale2 libupstart1 libva1
  linux-headers-3.13.0-43 linux-headers-3.13.0-43-generic
  linux-image-3.13.0-43-generic linux-image-extra-3.13.0-43-generic opus-tools
  python-beautifulsoup python-concurrent.futures python-docopt python-notify2
  python-protobuf python-psutil python-pycurl python-setproctitle sdparm
  vorbis-tools
Use 'apt-get autoremove' to remove them.
0 upgraded, 0 newly installed, 0 to remove and 44 not upgraded.

I16 encoder is still missing.

**Comment by masmu** [comment]
_sox_ is the l16 encoder. If you startup _pulseaudio-dlna_ there will be a line for the _l16_ encoder within the encoder settings. Like:
```
10-24 13:09:13 pulseaudio_dlna.application                    INFO     Encoder settings:
10-24 13:09:13 pulseaudio_dlna.application                    INFO       <AacEncoder available="True" bit-rate="192">
10-24 13:09:13 pulseaudio_dlna.application                    INFO       <FlacEncoder available="False">
10-24 13:09:13 pulseaudio_dlna.application                    INFO       <L16Encoder available="True" sample-rate="44100" channels="2">
```

If it is listed as `available="True"` everything is ok. E.g. you can see that i did not install the _flac_ encoder as it is listed as `available="False"`.

**Comment by JarlJansen** [comment]
Sorry, my mistake. It's L16 encoder, not I16 ;-) Tried to start with "--encoder I16" instead of "--encoder l16" ;-)

**Comment by JarlJansen** [comment]
Now it's like before using L16 encoder. Speed too fast, interruptions every 3 minutes.

**Comment by masmu** [comment]
Hmm, Ok. I placed some additional log in the branch. Please shutdown _pulseaudio-dlna_, do a `git pull` (will recieve my changes) restart it and please paste the output of it here. I assume it is still too fast?

**Comment by JarlJansen** [comment]
Think i should do another "make" before restart, right ?

**Comment by masmu** [comment]
Nope, this is required only once to make the application runnable.

**Comment by JarlJansen** [comment]
Sorry, still too fast.

http://pastebin.com/raw.php?i=rp2RVJbc

**Comment by masmu** [comment]
Ok, i resorted the sample rates. Please shutdown pulseaudio-dlna, do a `git pull` and restart it.

**Comment by JarlJansen** [comment]
Now as i am used to faster speed i would say "too slow" ;-) 

No, just a joke ! Works perfect !!! 
But i am still using "--encoder l16". Do I still need that ?

**Comment by masmu** [comment]
Glad to hear!

You can use the _devices.json_ which selects the encoder for you. I recommend to delete the folder we just created with the virtualenv. After that just install _pulseaudio-dlna_ via apt-get again and restore the _devices.json_. That reverts all changes. After that it should be as it was before we did some of the testing. You can then just start it via `pulseaudio-dlna` ...

Just for your information:
The HTTP header informes your _Xbox_ that the following data will be encoded with a specific sample rate. In your case _pulseaudio-dlna_ choosed 48khz because it was the highest settings available. The xbox (i assume) did not recognize that header and played it with 44.1khz and therefore it was faster then expected. Normally it should do as requested and play it with 48khz (as instructed through that HEADER) which would result in normal play speed. But somehow it ignores that.
So, my fix for that problem is to always prefer 44.1khz audio data, since the source data is encoded with that sample rate and there is no benefit in using a higher sample rate if the source quality is below that.

**Comment by masmu** [comment]
* __0.4.7__ - (_2015-11-18_)
    - The application can now co-exist with other applications which are using the port 1900/udp (thanks to [klaernie](https://github.com/klaernie))
    - Fixed the daemon mode to support `psutil` 1.x and 2.x (thanks to [klaernie](https://github.com/klaernie))
    - HTML entities in device descriptions are now converted automatically
    - Faster and more reliable device discovery (new dependency `python-netifaces`)
    - Added the `--cover-mode` option, one mode requires (optional) dependencies `gtk`, `cairo`, `rsvg`
    - L16 codecs are now selected better (e.g. needed for _XBox 360_)
    - Fixed a bug where sometimes it was tried to remove sinks twice on cleanup
    - Added the `--update-device-config` flag
    - Added the `--ssdp-ttl`, `--ssdp-mx`, `--ssdp-amount` options
    - Added the `--msearch-port` option
